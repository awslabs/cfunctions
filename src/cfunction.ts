import * as child_process from 'child_process';
import { createHash } from 'crypto';
import { tmpdir } from 'os';
import { join, resolve, dirname, relative } from 'path';
import { chdir } from 'process';

try {
  /* eslint-disable import/no-extraneous-dependencies */
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  require.resolve('esbuild/package.json');
  /* eslint-enable import/no-extraneous-dependencies */
} catch (e) {
  throw new Error('"esbuild" must be installed on the system');
}


// we are taking a "runtime peer" dependency on esbuild
// eslint-disable-next-line import/no-extraneous-dependencies
import * as esbuild from 'esbuild';

import { mkdtempSync, writeFileSync, readFileSync, copySync } from 'fs-extra';

export interface CFunctionProps {
  /**
   * Symbols to capture
   */
  readonly capture?: { [name: string]: any };

  /**
   * Javascript code to execute
   * @default "true;"
   */
  readonly code?: string;
}

export interface ExecOptions {
  /**
   * Environment variables to bind to the child process.
   *
   * You can use `cfunction.env` to bind the original symbols.
   *
   * @default {}
   */
  readonly env?: { [name: string]: string };
}

export class CFunction {
  public static exec(file: string, options: ExecOptions = {}): any {
    try {
      const substrate = `require('${file}')().catch(e => { console.error(e.stack); process.exit(1); }).then(o => console.log(JSON.stringify(o)))`;
      const child = child_process.spawnSync(process.execPath, ['-e', substrate], {
        stdio: ['inherit', 'pipe', 'pipe'],
        env: options.env ?? {},
      });

      const stderr = child.stderr.toString('utf-8');

      if (child.status !== 0) {
        throw new Error(`non-exit from child process: ${stderr}`);
      }

      // allow function to emit logs to STDERR (non-streaming)
      if (stderr.length) {
        console.error(stderr);
      }

      // STDOUT is the result
      return JSON.parse(child.stdout.toString());
    } catch (e) {
      throw new Error(`unable to execute ${file}: ${e.message}`);
    }
  }

  /**
   * The location of the function bundle (.js file).
   */
  public readonly outfile: string;

  /**
   * Environment variables that are expected to be available when the function
   * is executed.
   */
  public readonly env: { [name: string]: string };

  private readonly workdir: string;

  constructor(props: CFunctionProps) {
    const outdir = mkdtempSync(join(tmpdir(), '.cf.out.'));
    this.workdir = mkdtempSync(join(tmpdir(), '.cf.work.'));
    this.env = {};

    const workdir = this.workdir;

    // deploy runtime
    const runtimesrc = dirname(require.resolve(join(__dirname, '..', 'resources', 'cfruntime.js')));
    copySync(runtimesrc, workdir);

    const outfile = resolve(outdir, 'cf.js');

    const epcode = new Array<string>();

    this.captureSymbols(props.capture ?? { }, epcode);

    const code = props.code ?? 'true';
    epcode.push(`const output = ${code};`);
    epcode.push('module.exports = require(\'./cfruntime\').wrap(output);');

    const entrypoint = join(workdir, 'entrypoint.js');
    writeFileSync(entrypoint, epcode.join('\n'));

    const restore = process.cwd();
    chdir(workdir);
    try {
      const result = esbuild.buildSync({
        bundle: true,
        entryPoints: [relative(workdir, entrypoint)],
        outfile,
        platform: 'node',
        logLevel: 'silent',
      });
      if (result.warnings?.length) {
        throw new Error('esbuild warnings');
      }
    } catch (e) {
      throw new Error(`esbuild failed at ${workdir} ${e.message}`);
    } finally {
      chdir(restore);
    }

    this.outfile = outfile;
  }


  public toJson(): any {
    return {
      env: this.env,
      js: readFileSync(this.outfile, 'utf-8'),
    };
  }

  private captureSymbols(symbols: { [symbol: string]: any }, code: string[]) {
    for (const [name, value] of Object.entries(symbols)) {

      // if the value is a cfunction, we copy the cf.js file to our workdir and
      // bind the result of "require()"ing it locally.
      if (value instanceof CFunction) {
        const fileName = hashOf(value.outfile) + '.js';
        copySync(value.outfile, join(this.workdir, fileName));
        code.push(`const ${name} = require('./${fileName}');`);
        continue;
      }

      // bind the value through an environment variable (late binding). technically it should be
      // possible to do early-binding for all values that are not CDK tokens but since we are unable
      // to identify tokens at this layer, we have to deffer to late bindings.
      const cvalue = this.captureValue(value);
      const envVar = symbolToEnv(name);
      code.push(`const ${name} = process.env["${envVar}"];`);
      code.push(`if (${name} === undefined) { throw new Error("missing required environment variable '${envVar}'"); }`);

      this.env[envVar] = cvalue;
    }
  }

  private captureValue(x: any) {
    if (typeof(x) === 'function') {
      throw new Error('cannot capture functions yet');
    };
    if (typeof(x) === 'symbol') {
      throw new Error('cannot capture symbols yet');
    }

    if (Array.isArray(x)) {
      return this.captureArray(x);
    }

    if (typeof(x) === 'object') {
      return this.captureObject(x);
    }

    return this.capturePrimitive(x);
  }

  private capturePrimitive(x: string | number | boolean | bigint) {
    return JSON.stringify(x);
  }

  private captureObject(o: any) {
    if (o.constructor.name != 'Object') {
      throw new Error('classes are not supported yet');
    }

    if ('toJson' in o && typeof(o.toJson === 'function')) {
      o = o.toJson();
    }

    const fields = Array<string>();
    for (const [k, v] of Object.entries(o)) {
      fields.push(`"${k}": ${this.captureValue(v)}`);
    }

    return `{ ${fields.join(',' )}}`;
  }

  private captureArray(a: any[]): string {
    return '[' + a.map(x => this.captureValue(x)).join(',') + ']';
  }
}

/** generate an environment variable for a named symbol */
function symbolToEnv(symbol: string) {
  return `__CF__${symbol}__`;
}


function hashOf(file: string): string {
  const sha1 = createHash('sha1');
  sha1.update(readFileSync(file));
  return sha1.digest('hex');
}
