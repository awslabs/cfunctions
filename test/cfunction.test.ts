import { CFunction, CFunctionProps } from '../src';
import { snap, fails } from './util';

test('no capture', snap(() => new CFunction({
  capture: {},
  code: 'true;',
})));

test('compilation error', fails(() => new CFunction({
  code: 'i..;',
}), /esbuild failed/));

test('primitive', snap(() => new CFunction({
  capture: { hello: 22 },
  code: 'true;',
})));

test('array objects', snap(() => new CFunction({
  capture: { hello: [1, 2, 'hello', true, { world: 1 }] },
  code: 'true;',
})));

test('with classees', fails(() => {
  class Hello { }

  return new CFunction({
    capture: { a_class: new Hello() },
    code: 'true;',
  });
}, /classes are not supported/));

test('multiple captures', snap(() => new CFunction({
  capture: {
    hello: 'yes',
    data: {
      field1: ['a', 'b'],
      field2: true,
    },
  },
})));

test('respects toJson()', snap(() => new CFunction({
  capture: {
    hello: { toJson: () => 14 },
  },
})));

test('rejects functions', fails(() => new CFunction({
  capture: {
    boom: () => 123,
  },
}), /cannot capture functions/));

test('a real example', snap(() => new CFunction({
  capture: {
    x: 12,
    y: 78,
  },
  code: 'x + y',
})));

test('output is executable', snap(() => {
  const cf = new CFunction({
    capture: { lhs: 10, rhs: 30 },
    code: 'lhs * rhs',
  });

  const result = CFunction.exec(cf.outfile, {
    env: cf.env,
  });

  expect(result).toBe(300);

  return result;
}));

test('exec without required variables', fails(() => {
  const cf = new CFunction({
    capture: { lhs: 10, rhs: 30 },
    code: 'lhs * rhs',
  });

  CFunction.exec(cf.outfile);
}, /missing required environment variable '__CF__lhs__'/));

function captureAndExecute(props: CFunctionProps) {
  const cf = new CFunction(props);
  return CFunction.exec(cf.outfile);
}

test('function literal', snap(() => {
  const result = captureAndExecute({
    code: (() => { return 'value from a function'; }).toString(),
  });

  expect(result).toBe('value from a function');

  return result;
}));

test('async literal', snap(() => {
  const result = captureAndExecute({
    code: (async () => { return 'value from a function'; }).toString(),
  });

  expect(result).toStrictEqual('value from a function');

  return result;
}));

test('function', snap(() => {
  function code() {
    return 42;
  }

  const result = captureAndExecute({
    code: code.toString(),
  });

  expect(result).toBe(42);
  return result;
}));

test('async function', snap(() => {
  async function code() {
    async function f2(x: number) {
      return x * 2;
    }

    return {
      yes: 'async',
      no: await f2(50),
    };
  }

  const result = captureAndExecute({
    code: code.toString(),
  });

  return result;
}));

test('capture another cfunction', snap(() => {
  const f1 = new CFunction({
    code: '() => 42',
  });

  const f2 = new CFunction({
    code: 'async function(x) { return 3 * x; }',
  });

  const f3 = new CFunction({
    capture: {
      f1,
      boom: f2,
    },
    code: 'async () => await f1() + 2 + await boom(77)',
  });

  const result = CFunction.exec(f3.outfile);
  expect(result).toStrictEqual(42 + 2 + 3 * 77);

  return f3;
}));
