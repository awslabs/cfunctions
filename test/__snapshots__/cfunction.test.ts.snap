// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a real example 1`] = `
Object {
  "env": Object {
    "__CF__x__": "12",
    "__CF__y__": "78",
  },
  "js": "var __commonJS = (callback, module2) => () => {
  if (!module2) {
    module2 = {exports: {}};
    callback(module2.exports, module2);
  }
  return module2.exports;
};

// cfruntime.js
var require_cfruntime = __commonJS((exports2, module2) => {
  function wrap(output2) {
    return async function(...args) {
      if (typeof output2 === \\"function\\") {
        output2 = output2(...args);
      }
      if (isPromise(output2)) {
        output2 = await output2;
      }
      return output2;
    };
  }
  function isPromise(x2) {
    return typeof x2 === \\"object\\" && \\"then\\" in x2 && \\"catch\\" in x2;
  }
  module2.exports = {wrap};
});

// entrypoint.js
var x = process.env[\\"__CF__x__\\"];
if (x === void 0) {
  throw new Error(\\"missing required environment variable '__CF__x__'\\");
}
var y = process.env[\\"__CF__y__\\"];
if (y === void 0) {
  throw new Error(\\"missing required environment variable '__CF__y__'\\");
}
var output = x + y;
module.exports = require_cfruntime().wrap(output);
",
}
`;

exports[`array objects 1`] = `
Object {
  "env": Object {
    "__CF__hello__": "[1,2,\\"hello\\",true,{ \\"world\\": 1}]",
  },
  "js": "var __commonJS = (callback, module2) => () => {
  if (!module2) {
    module2 = {exports: {}};
    callback(module2.exports, module2);
  }
  return module2.exports;
};

// cfruntime.js
var require_cfruntime = __commonJS((exports2, module2) => {
  function wrap(output2) {
    return async function(...args) {
      if (typeof output2 === \\"function\\") {
        output2 = output2(...args);
      }
      if (isPromise(output2)) {
        output2 = await output2;
      }
      return output2;
    };
  }
  function isPromise(x) {
    return typeof x === \\"object\\" && \\"then\\" in x && \\"catch\\" in x;
  }
  module2.exports = {wrap};
});

// entrypoint.js
var hello = process.env[\\"__CF__hello__\\"];
if (hello === void 0) {
  throw new Error(\\"missing required environment variable '__CF__hello__'\\");
}
var output = true;
module.exports = require_cfruntime().wrap(output);
",
}
`;

exports[`async function 1`] = `
Object {
  "no": 100,
  "yes": "async",
}
`;

exports[`async literal 1`] = `"value from a function"`;

exports[`capture another cfunction 1`] = `
Object {
  "env": Object {},
  "js": "var __commonJS = (callback, module2) => () => {
  if (!module2) {
    module2 = {exports: {}};
    callback(module2.exports, module2);
  }
  return module2.exports;
};

// a7db9f5b39aff35858c21324987f4635dae174e5.js
var require_a7db9f5b39aff35858c21324987f4635dae174e5 = __commonJS((exports2, module2) => {
  var __commonJS2 = (callback, module22) => () => {
    if (!module22) {
      module22 = {exports: {}};
      callback(module22.exports, module22);
    }
    return module22.exports;
  };
  var require_cfruntime2 = __commonJS2((exports22, module22) => {
    function wrap(output22) {
      return async function(...args) {
        if (typeof output22 === \\"function\\") {
          output22 = output22(...args);
        }
        if (isPromise(output22)) {
          output22 = await output22;
        }
        return output22;
      };
    }
    function isPromise(x) {
      return typeof x === \\"object\\" && \\"then\\" in x && \\"catch\\" in x;
    }
    module22.exports = {wrap};
  });
  var output2 = () => 42;
  module2.exports = require_cfruntime2().wrap(output2);
});

// 8c53027a78125d55c36a1d2e8a01c52e2fcb772b.js
var require_c53027a78125d55c36a1d2e8a01c52e2fcb772b = __commonJS((exports2, module2) => {
  var __commonJS2 = (callback, module22) => () => {
    if (!module22) {
      module22 = {exports: {}};
      callback(module22.exports, module22);
    }
    return module22.exports;
  };
  var require_cfruntime2 = __commonJS2((exports22, module22) => {
    function wrap(output22) {
      return async function(...args) {
        if (typeof output22 === \\"function\\") {
          output22 = output22(...args);
        }
        if (isPromise(output22)) {
          output22 = await output22;
        }
        return output22;
      };
    }
    function isPromise(x) {
      return typeof x === \\"object\\" && \\"then\\" in x && \\"catch\\" in x;
    }
    module22.exports = {wrap};
  });
  var output2 = async function(x) {
    return 3 * x;
  };
  module2.exports = require_cfruntime2().wrap(output2);
});

// cfruntime.js
var require_cfruntime = __commonJS((exports2, module2) => {
  function wrap(output2) {
    return async function(...args) {
      if (typeof output2 === \\"function\\") {
        output2 = output2(...args);
      }
      if (isPromise(output2)) {
        output2 = await output2;
      }
      return output2;
    };
  }
  function isPromise(x) {
    return typeof x === \\"object\\" && \\"then\\" in x && \\"catch\\" in x;
  }
  module2.exports = {wrap};
});

// entrypoint.js
var f1 = require_a7db9f5b39aff35858c21324987f4635dae174e5();
var boom = require_c53027a78125d55c36a1d2e8a01c52e2fcb772b();
var output = async () => await f1() + 2 + await boom(77);
module.exports = require_cfruntime().wrap(output);
",
}
`;

exports[`function 1`] = `42`;

exports[`function literal 1`] = `"value from a function"`;

exports[`multiple captures 1`] = `
Object {
  "env": Object {
    "__CF__data__": "{ \\"field1\\": [\\"a\\",\\"b\\"],\\"field2\\": true}",
    "__CF__hello__": "\\"yes\\"",
  },
  "js": "var __commonJS = (callback, module2) => () => {
  if (!module2) {
    module2 = {exports: {}};
    callback(module2.exports, module2);
  }
  return module2.exports;
};

// cfruntime.js
var require_cfruntime = __commonJS((exports2, module2) => {
  function wrap(output2) {
    return async function(...args) {
      if (typeof output2 === \\"function\\") {
        output2 = output2(...args);
      }
      if (isPromise(output2)) {
        output2 = await output2;
      }
      return output2;
    };
  }
  function isPromise(x) {
    return typeof x === \\"object\\" && \\"then\\" in x && \\"catch\\" in x;
  }
  module2.exports = {wrap};
});

// entrypoint.js
var hello = process.env[\\"__CF__hello__\\"];
if (hello === void 0) {
  throw new Error(\\"missing required environment variable '__CF__hello__'\\");
}
var data = process.env[\\"__CF__data__\\"];
if (data === void 0) {
  throw new Error(\\"missing required environment variable '__CF__data__'\\");
}
var output = true;
module.exports = require_cfruntime().wrap(output);
",
}
`;

exports[`no capture 1`] = `
Object {
  "env": Object {},
  "js": "var __commonJS = (callback, module2) => () => {
  if (!module2) {
    module2 = {exports: {}};
    callback(module2.exports, module2);
  }
  return module2.exports;
};

// cfruntime.js
var require_cfruntime = __commonJS((exports2, module2) => {
  function wrap(output2) {
    return async function(...args) {
      if (typeof output2 === \\"function\\") {
        output2 = output2(...args);
      }
      if (isPromise(output2)) {
        output2 = await output2;
      }
      return output2;
    };
  }
  function isPromise(x) {
    return typeof x === \\"object\\" && \\"then\\" in x && \\"catch\\" in x;
  }
  module2.exports = {wrap};
});

// entrypoint.js
var output = true;
module.exports = require_cfruntime().wrap(output);
",
}
`;

exports[`output is executable 1`] = `300`;

exports[`primitive 1`] = `
Object {
  "env": Object {
    "__CF__hello__": "22",
  },
  "js": "var __commonJS = (callback, module2) => () => {
  if (!module2) {
    module2 = {exports: {}};
    callback(module2.exports, module2);
  }
  return module2.exports;
};

// cfruntime.js
var require_cfruntime = __commonJS((exports2, module2) => {
  function wrap(output2) {
    return async function(...args) {
      if (typeof output2 === \\"function\\") {
        output2 = output2(...args);
      }
      if (isPromise(output2)) {
        output2 = await output2;
      }
      return output2;
    };
  }
  function isPromise(x) {
    return typeof x === \\"object\\" && \\"then\\" in x && \\"catch\\" in x;
  }
  module2.exports = {wrap};
});

// entrypoint.js
var hello = process.env[\\"__CF__hello__\\"];
if (hello === void 0) {
  throw new Error(\\"missing required environment variable '__CF__hello__'\\");
}
var output = true;
module.exports = require_cfruntime().wrap(output);
",
}
`;

exports[`respects toJson() 1`] = `
Object {
  "env": Object {
    "__CF__hello__": "{ }",
  },
  "js": "var __commonJS = (callback, module2) => () => {
  if (!module2) {
    module2 = {exports: {}};
    callback(module2.exports, module2);
  }
  return module2.exports;
};

// cfruntime.js
var require_cfruntime = __commonJS((exports2, module2) => {
  function wrap(output2) {
    return async function(...args) {
      if (typeof output2 === \\"function\\") {
        output2 = output2(...args);
      }
      if (isPromise(output2)) {
        output2 = await output2;
      }
      return output2;
    };
  }
  function isPromise(x) {
    return typeof x === \\"object\\" && \\"then\\" in x && \\"catch\\" in x;
  }
  module2.exports = {wrap};
});

// entrypoint.js
var hello = process.env[\\"__CF__hello__\\"];
if (hello === void 0) {
  throw new Error(\\"missing required environment variable '__CF__hello__'\\");
}
var output = true;
module.exports = require_cfruntime().wrap(output);
",
}
`;
